### PR TITLE
Enable full test suite with Python 3.

### PR DESCRIPTION
**Pull request by jmchilton**
Tweaking testing in a way that needs to be stripped out before merge but hoping to get some fresh Python 3 tracebacks on Travis to work from.

Problems:
- Something in dependency_script
   ```Error processing /home/travis/build/galaxyproject/planemo/tests/data/repos/package_1/tool_dependencies.xml - Checksum failure for /tmp/tmph9zwo1rp/download_cache/samtools-0.1.16-linux-i386.tgz, got b'7090bd62142df0ed3a2b4000292fc0edc917e773d0269c2f3f5b0d36b7888c22' but wanted '7090bd62142df0ed3a2b4000292fc0edc917e773d0269c2f3f5b0d36b7888c22'```
- Galaxy getting loaded without target tool in the tool panel - other people have observed this in the wild - seems now somehow related to Python 3 Planemo?
-
 ```======================================================================
  FAIL: test_serve_workflow (tests.test_cmd_serve.ServeTestCase)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/home/travis/build/galaxyproject/planemo/tests/test_cmd_serve.py", line 54, in 
  test_serve_workflow
      assert len(user_gi.workflows.get_workflows()) == 1
  nose.proxy.AssertionError: 
      assert len(GalaxyInstance object for Galaxy at 
  http://localhost:51267.histories.get_histories(name='Cool History 42')) == 0
      GalaxyInstance object for Galaxy at http://localhost:51267.histories.create_history('Cool History 
  42')
      assert GalaxyInstance object for Galaxy at 
  http://localhost:51267.tools.get_tools(tool_id="random_lines1")
  >>  assert len(GalaxyInstance object for Galaxy at http://localhost:51267.workflows.get_workflows()) == 1
```
  Probably same problem as above?

### CODE REVIEW

**Comment by nsoranzo** [comment]
Commit https://github.com/galaxyproject/planemo/pull/774/commits/416a332be93e93de241f5d2aceadb51ebaac0788 fixed the last point on the PR description, which I have now removed. Now tests are able to download the latest SQLite database and therefore run faster.

**Comment by nsoranzo** [comment]
Down to 4 failures, I may need some help with those.

**Comment by jmchilton** [comment]
Wow - thanks so much @nsoranzo - this is awesome.

Trying to pick it up from here - I can replicate broadly the behavior here https://github.com/galaxyproject/planemo/issues/765 with Python 3 now - the toolbox loads fine with Python 3 and ``planemo serve`` but there are no tests with ``planemo test``. This is at least one of the 4 remaining errors - this one about expecting a test failure and not finding one.

**Comment by nsoranzo** [comment]
Thanks! I tried to restart the test build a few times, but it keeps stalling at "Test testing a simple GA workflow with Galaxy" test.

**Comment by jmchilton** [comment]
Working on it - locks up pretty quickly here. Looks like the output is fetched from Galaxy but nothing ends.

```
galaxy.jobs DEBUG 2018-02-26 21:07:34,550 job 3 ended (finish() executed in (1415.952 ms))
127.0.0.1 - - [26/Feb/2018:21:07:34 -0400] "GET /api/workflows/2891970512fa2d5a/invocations/2891970512fa2d5a?key=a293420304dd548281d311bc6ff1f0ba HTTP/1.1" 200 - "-" "python-requests/2.18.4"
127.0.0.1 - - [26/Feb/2018:21:07:34 -0400] "GET /api/jobs/54f2a3a23292eb07?full=True&key=a293420304dd548281d311bc6ff1f0ba HTTP/1.1" 200 - "-" "python-requests/2.18.4"
127.0.0.1 - - [26/Feb/2018:21:07:35 -0400] "GET /api/histories/2891970512fa2d5a/contents/54f2a3a23292eb07?key=a293420304dd548281d311bc6ff1f0ba HTTP/1.1" 200 - "-" "python-requests/2.18.4"
127.0.0.1 - - [26/Feb/2018:21:07:35 -0400] "GET /api/histories/2891970512fa2d5a/contents/54f2a3a23292eb07/display HTTP/1.1" 200 - "-" "python-requests/2.18.4"
```

Maybe killing the server and re-joining is broken in Python 3... hmm...

**Comment by jmchilton** [comment]
This piece of code in galaxy-lib I suspect is the problem:

```python
    if path is not None:
        properties["path"] = path
        f = open(path, "rb")
    else:
        f = StringIO(content)

    try:
        contents = f.read(1024 * 1024)
        filesize = 0
        while contents != "":
            checksum.update(contents)
            filesize += len(contents)
            contents = f.read(1024 * 1024)
    finally:
        f.close()
```

**Comment by peterjc** [comment]
@jmchilton Given the file is opened in binary mode, and you want the raw length, then ``from io import BytesIO`` seems a better bet than ``StringIO`` here.

**Comment by jmchilton** [comment]
@peterjc Yup - that was exactly the fix - https://github.com/galaxyproject/galaxy-lib/pull/93. I should have linked it against this PR.

**Comment by jmchilton** [comment]
I think the only broken test here is a test that has been broken in master as we worked on this. Local conda packages maybe broke during the big galaxy-lib update or maybe broke due to Conda changes elsewhere.

File: planemo/commands/cmd_project_init.py [hunk 1]
```diff
@@ -43,11 +43,11 @@ def cli(ctx, path, template=None, **kwds):
         untar_args = UNTAR_ARGS % (tempdir)
         untar_to(DOWNLOAD_URL, tempdir, untar_args)
         template_dir = os.path.join(tempdir, template)
-        shell("ls '%s'" % (template_dir))
-        shell("mv '%s'/* '%s'" % (template_dir, path))
+        shell(['ls', template_dir])
+        shell(['mv', "%s/*" % template_dir, path])
```
Comment (jmchilton) [inline planemo/commands/cmd_project_init.py]: Thanks for the fix - the rest of this looks good but this line required the shell expansion to happen. Can I revert just this line or are you working on an approach that avoids shell to do this?

Comment (nsoranzo) [reply planemo/commands/cmd_project_init.py]: Yep, this broke project_init tests. Rebased with a fixed (and shorter) version.

**Comment by jmchilton** [comment]
The jmchilton branch Travis finishes faster:

```
___________________________________ summary ____________________________________
  py34: commands succeeded
  congratulations :)
```

🎆 🎆 🎆 🎆 🎆 

Thanks so much @nsoranzo - this is really awesome. I really appreciate all the work you put into this.

**Pull request merged**